### PR TITLE
Use doctrine instantiator

### PR DESCRIPTION
**Pull request by gsomoza**
See https://github.com/zfcampus/zf-apigility-doctrine/pull/271

### CODE REVIEW

**Comment by gsomoza** [comment]
@webimpress FYI

File: src/Server/Resource/DoctrineResourceFactory.php [hunk 1]
```diff
@@ -99,14 +99,17 @@ public function __invoke(ContainerInterface $container, $requestedName, array $o
 
         $resourceClass = $this->getResourceClassFromConfig($doctrineConnectedConfig, $requestedName);
         $objectManager = $container->get($doctrineConnectedConfig['object_manager']);
+        $entityFactory = ! empty($doctrineConnectedConfig['entity_factory'])
```
Comment (TomHAnderson) [inline src/Server/Resource/DoctrineResourceFactory.php]: Change to isset or provide reason for empty

Comment (gsomoza) [inline src/Server/Resource/DoctrineResourceFactory.php]: `!empty` because it could be set to `''` and we also don't want that.

File: test/src/Server/ODM/CRUD/CRUDTest.php [hunk 2]
```diff
@@ -129,6 +132,46 @@ function (DoctrineResourceEvent $e) {
...
+    /**
+     * @return void
+     */
+    public function testCreateByExplicitlySettingEntityFactoryInConstructor()
+    {
+        /** @var InstantiatorInterface|\PHPUnit_Framework_MockObject_MockObject $entityFactoryMock */
+        $entityFactoryMock = $this->getMockBuilder(InstantiatorInterface::class)->getMock();
+        $entityFactoryMock->expects(self::once())
+            ->method('instantiate')
+            ->with('ZFTestApigilityDbMongo\Document\Meta')
```
Comment (webimpress) [inline test/src/Server/ODM/CRUD/CRUDTest.php]: Please use `::class` notation.

File: test/src/Server/ODM/CRUD/CRUDTest.php [hunk 3]
```diff
@@ -129,6 +132,46 @@ function (DoctrineResourceEvent $e) {
...
+
+
+        /** @var ServiceManager $sm */
+        $sm = $this->getApplication()->getServiceManager();
+
+        $config = $sm->get('config');
+        $resourceName = 'ZFTestApigilityDbMongoApi\V1\Rest\Meta\MetaResource';
+        $resourceConfig = ! empty($config['zf-apigility']['doctrine-connected'][$resourceName])
+            ? $config['zf-apigility']['doctrine-connected'][$resourceName]
+            : [];
```
Comment (webimpress) [inline test/src/Server/ODM/CRUD/CRUDTest.php]: Hm, why do we need it here? This is test, so we should know exactly what is our input data.

File: test/src/Server/ODM/CRUD/CRUDTest.php [hunk 4]
```diff
@@ -129,6 +132,46 @@ function (DoctrineResourceEvent $e) {
...
+        $resourceConfig['entity_factory'] = 'ResourceInstantiator';
+        $config['zf-apigility']['doctrine-connected'][$resourceName] = $resourceConfig;
+
+        $sm->setAllowOverride(true);
+        $sm->setService('config', $config);
+        $sm->setAllowOverride(false);
+
+        $sm->setFactory(
```
Comment (webimpress) [inline test/src/Server/ODM/CRUD/CRUDTest.php]: Hm. Can we use `setService` here? We already have instance of the entity factory.

File: test/src/Server/ODM/CRUD/CRUDTest.php [hunk 5]
```diff
@@ -129,6 +132,46 @@ function (DoctrineResourceEvent $e) {
...
+            'ResourceInstantiator',
+            function () use ($entityFactoryMock) {
+                return $entityFactoryMock;
+            }
+        );
+
+        $this->testCreate();
```
Comment (webimpress) [inline test/src/Server/ODM/CRUD/CRUDTest.php]: I don't like this type of testing, it is very unclear what's going on there. Sometimes we use `@depends` annotation, and it's fine, because we can see there dependency between tests.

File: test/src/Server/ODM/CRUD/CRUDTest.php [hunk 6]
```diff
@@ -427,7 +470,7 @@ protected function validateTriggeredEvents(array $expectedEvents)
     protected function createMeta($name = null)
     {
         $meta = new Meta();
-        $meta->setName($name ?: 'Meta Name');
+        $meta->setName($name ? : 'Meta Name');
```
Comment (webimpress) [inline test/src/Server/ODM/CRUD/CRUDTest.php]: Why this change? Please revert.

**Comment by webimpress** [comment]
@gsomoza this is feature, so PR should be against develop branch. I've changed it for the PR, please rebase against develop. Thanks!

**Comment by gsomoza** [comment]
Still have to polish the tests after the rebase, will do that as soon as I get a chance (maybe tonight).

File: test/src/Server/Event/Listener/CollectionListenerTest.php [hunk 7]
```diff
@@ -0,0 +1,80 @@
+<?php
+/**
+ * @license   http://opensource.org/licenses/BSD-3-Clause BSD-3-Clause
+ * @copyright Copyright (c) 2016 Zend Technologies USA Inc. (http://www.zend.com)
```
Comment (webimpress) [inline test/src/Server/Event/Listener/CollectionListenerTest.php]: As it is new file, please change year here to `2017`.

File: src/Server/Resource/DoctrineResource.php [hunk 8]
```diff
@@ -86,6 +87,17 @@ class DoctrineResource extends AbstractResourceListener implements
      */
     protected $hydrator;
 
+    /** @var InstantiatorInterface|null */
```
Comment (webimpress) [inline src/Server/Resource/DoctrineResource.php]: As I mention before, please change it to 3-line comment, to be consistent in the file:

```php
/**
 * @var InstantiatorInterface|null
 */
```

File: src/Server/Resource/DoctrineResource.php [hunk 9]
```diff
@@ -330,7 +344,10 @@ public function create($data)
             return $data;
         }
 
-        $entity = new $entityClass;
+        $entity = $this->entityFactory
+            ? $this->entityFactory->instantiate($entityClass)
```
Comment (alextech) [inline src/Server/Resource/DoctrineResource.php]: Does Instantiator interface really work here, since construction could depend on $data? Wouldn't also need to pass $data to factory to make this useful?

**Comment by alextech** [comment]
Found this PR as I was struggling with same problem.  Entities are domain definitions and doctrine is just ORM utility connecting it to storage and API, so it should not dictate how my entities are instantiated at domain level.  Therefore, if entity has some constructor requirements to avoid being in invalid state, factory or instantiator need to be passed full input data.

Uploaded data can be a mixture of required and optional parameters, so:

- Factory should create entity instance based on required subset of keys, `$entity = $this->entityFactory->instantiate($entityClass, $data)`.  It will know how to use input data to create entity.
- Hydrators pick up the rest of optional keys.

Problem I am seeing is `InstantiatorInterface::instantiate` only takes in class name, not data.

File: src/Server/Resource/DoctrineResource.php (hunk 9 above)
Comment (gsomoza) [reply src/Server/Resource/DoctrineResource.php]: I know what you mean, but this PR is at least a simple, backwards-compatible first step. In fact I thought this PR was merged already and I'm surprised it still isn't...

**Comment by gsomoza** [comment]
@alextech I would recommend separating the Doctrine layer from your domain layer. In other words your Doctrine entities (DE) would be considered infrastructure in your domain model. You could make it so that each DE "knows" how to instantiate a domain entity. E.g.:

```php
namespace \Acme\Billing;
class Invoice {} // your domain entity / value object

namespace \Acme\Doctrine\Billing {}
class InvoiceEntity {
    public function toInvoice(): \Acme\Billing\Invoice {
        return new \Acme\Billing\Invoice($this->getId(), $this->getItems() /* etc */);
    }
}
```
This would mean the `InvoiceEntity` can be instantiated without a constructor because it doesn't represent a valuable business object like an `Invoice` - it just represents an item (row/document) in a database.

**Comment by gsomoza** [comment]
I think the answer to your questions comes down to how Apigility is being used together with Doctrine:

A) To provide a **DBAL API**

The fastest approach because it's just mapping to ORM to API endpoints. In this case, the Doctrine Entity is very close to the "surface" of your API. E.g. you'd be using `DoctrineResource` to expose Doctrine Entities almost 1:1 as API endpoints, with some goodies on top like validation, filtering, hydrators, pagination, etc. In other words, you'd just be allowing your API clients to communicate remotely with your database using REST (as opposed to communicating locally using an ORM).
* This is the use-case that has extensive documentation in Apigility.
* You can build on top of it to apply some domain logic through e.g. value objects consumed by your Doctrine entities, but even that has it's own "gotchas" and you'd have to understand the internals of Doctrine very well to leverage VO's properly.
* Best for small/simple APIs that don't require much business logic. E.g. some internal services or APIs to power a special admin interface, etc.

 B) To provide a **Business API** (maybe there's a better way to name this?)

Endpoints here would represent resources that send messages to your business model (for "writes") and read data from the infrastructure (for "reads"). 
* Writes: inputs are filtered, validated, etc., by Apigility. Then they're mapped to a "message" sent to your domain model (e.g. a "command" if you're using CQRS). I.e. Request > Apigility > "Domain Resource" (AFAIK you'd have to build this yourself) > Domain Command > Domain Model. 
* Reads: data is read directly from the infrastructure in most cases – e.g. from projection tables. I.e. Request > Apigility > Doctrine Entity for Projection > Response.

C) There are other setups that weigh heavier in one or the other direction, or in completely different directions, but the two above are the major ones I've worked with in the past years.

Keeping in mind that I know very little I know about you or your project: it seems like you may have started building "A" (by following the documentation), but you actually needed "B". To get from "A" to "B" you'd have to replace your `DoctrineResource` API endpoints and instead use "custom code" endpoints that leverage your domain model.

I hope this helps you one way or the other.

**Comment by alextech** [comment]
>you'd have to replace your DoctrineResource API endpoints and instead use "custom code" 

Interesting you mention that, I started doing that first with just `EntityResource`, noticed it looked too much like reusable boilerplate, specifically where I had to bring in hydrators after entity instance creation, fetch existing entity if updating, then persist + flush cycle. So I looked around, found this repository, which helped with hydrator fetching and a really nice feature where it looks for presence of identifier field in payload to decide if entity needs creating or updating; if needs updating, performs `find` from `entityManager`. But it seemed incomplete. 

If I think hard about your (B) point, and how it is intended to be used as "projection of tables" then I guess your approach is not wrong, it just does not fit my style.  I guess I think about my REST API endpoints too much like RPC calls, not as table gateways.

**Comment by gsomoza** [comment]
Great, I'm glad if anything I said helped at all 👍

**Comment by webimpress** [comment]
Thanks, @gsomoza !
